Signal/src/ViewControllers/ConversationView/ConversationViewModel.m

```diff
@@ -0,0 +1,1284 @@
+    for (NSString *itemId in insertedItemIdList) {
+        OWSAssertDebug(![oldItemIdSet containsObject:itemId]);
+        OWSAssertDebug([newItemIdSet containsObject:itemId]);
+
+        NSUInteger newIndex = [newItemIdList indexOfObject:itemId];
+        if (newIndex == NSNotFound) {
+            OWSFailDebug(@"Can't find index of inserted view item.");
+            return [self.delegate conversationViewModelDidUpdate:ConversationUpdate.reloadUpdate];
+        }
+        id<ConversationViewItem> _Nullable viewItem = newViewItemMap[itemId];
+        if (!viewItem) {
+            OWSFailDebug(@"Can't find inserted view item.");
+            return [self.delegate conversationViewModelDidUpdate:ConversationUpdate.reloadUpdate];
+        }
+
+        [updateItems addObject:[[ConversationUpdateItem alloc] initWithUpdateItemType:ConversationUpdateItemType_Insert
+                                                                             oldIndex:NSNotFound
+                                                                             newIndex:newIndex
+                                                                             viewItem:viewItem]];
+        [transformedItemList insertObject:itemId atIndex:newIndex];
+    }
+
+    if (![newItemIdList isEqualToArray:transformedItemList]) {
+        // We should be able to represent all transformations as a series of
+        // inserts, updates and deletes - moves should not be necessary.
+        //
+        // TODO: The unread indicator might end up being an exception.
+        OWSLogWarn(@"New and updated view item lists don't match.");
+        return [self.delegate conversationViewModelDidUpdate:ConversationUpdate.reloadUpdate];
+    }
+
+    // 3. Updates.
+    //
+    // NOTE: Order doesn't matter.
+    for (NSString *itemId in updatedItemSet) {
+        if (![newItemIdList containsObject:itemId]) {
+            OWSFailDebug(@"Updated view item not in new view item list.");
+            continue;
+        }
+        if ([insertedItemIdList containsObject:itemId]) {
+            continue;
+        }
+        NSUInteger oldIndex = [oldItemIdList indexOfObject:itemId];
+        if (oldIndex == NSNotFound) {
+            OWSFailDebug(@"Can't find old index of updated view item.");
+            return [self.delegate conversationViewModelDidUpdate:ConversationUpdate.reloadUpdate];
+        }
+        NSUInteger newIndex = [newItemIdList indexOfObject:itemId];
+        if (newIndex == NSNotFound) {
+            OWSFailDebug(@"Can't find new index of updated view item.");
+            return [self.delegate conversationViewModelDidUpdate:ConversationUpdate.reloadUpdate];
+        }
+        id<ConversationViewItem> _Nullable viewItem = newViewItemMap[itemId];
+        if (!viewItem) {
+            OWSFailDebug(@"Can't find inserted view item.");
+            return [self.delegate conversationViewModelDidUpdate:ConversationUpdate.reloadUpdate];
+        }
```

Comment:
We're super defensive when constructing these updates because of the bugs we've seen in the YDB mappings.  One of the benefits to the CVM is that any inconsistencies/bugs in the YDB state/logic are _discarded_ by the CVM.  The CVC only knows about the CVM state (a sequence of items) and learns about changes to that state.  If YDB gets into a bad state, we might omit certain items from the CVM state, but that's hidden from the POV of the CVC.

Reply:
👍